File: kadai_3.py
#coding: utf-8

f =open('array.txt')
data = f.read().replace('\n','') #改行を無視しながら
f.close()


EcoRI = data.split('GAATTC') #EcoRIで切断、リスト化

print 'EcoRI\n断片数は'+str(len(EcoRI))+'本' #6個(要素数6のリスト)
i=-1
for j in EcoRI:
	i += 1 
	if i==0:
		print str(len(EcoRI[i])+1) +'bp' #最初　Gを加える
		
	elif i!= len(EcoRI)-1:
		print str(len(EcoRI[i])+6) +'bp' #最初最後以外　AATTCとGを加える
	else:
		print str(len(EcoRI[i])+5) +'bp\n' #最後 AATTCを加える


Hind = data.split('AAGCTT')
print 'HindIII\n断片数は'+str(len(Hind))+'本' 
i=-1
for j in Hind:
	i += 1 
	if i==0:
		print str(len(Hind[i])+1) +'bp' 
		
	elif i!= len(Hind)-1:
		print str(len(Hind[i])+6) +'bp' 
	else:
		print str(len(Hind[i])+5) +'bp\n' 


BamHI = data.split('GGATCC')
print 'BamHI\n断片数は'+str(len(BamHI))+'本' 
i=-1
for j in BamHI:
	i += 1 
	if i==0:
		print str(len(BamHI[i])+1) +'bp' 
		
	elif i!= len(BamHI)-1:
		print str(len(BamHI[i])+6) +'bp' 
	else:
		print str(len(BamHI[i])+5) +'bp\n' 

NotI = data.split('GCGGCCGC')
print 'BamHI\n断片数は'+str(len(NotI))+'本' 
i=-1
for j in NotI:
	i += 1 
	if i==0:
		print str(len(NotI[i])+1) +'bp' 
		
	elif i!= len(NotI)-1:
		print str(len(NotI[i])+6) +'bp' 
	else:
		print str(len(NotI[i])+5) +'bp\n' 

import re #区切り文字を複数使用するためにreモジュール
All = re.split('GAATTC|AAGCTT|GGATCC|GCGGCCGC',data)
print '全部\n断片数は'+str(len(All))+'本' 
i=-1
for j in All:
	i += 1 
	if i==0:
		print str(len(All[i])+1) +'bp' 
		
	elif i!= len(All)-1:
		print str(len(All[i])+6) +'bp' 
	else:
		print str(len(All[i])+5) +'bp' 

File: kadai_3.c
#include<stdio.h>
#include<string.h>

#define base 50000 //塩基数

int main(void)
{
	FILE *fp;
	int k;
	int i=0,j=0,c=0,d=0;
	char all[base]; //最初に全て文字を読み込む配列
	char data[base]; //改行なし(ATGCのみ)の配列
	char l[base]; //断片長の判断用

	fp=fopen("array.txt","r"); 
	
	if(fp == NULL){
		printf("ファイルをオープンできませんでした。\n");
		return 1;
	}
	else{
		printf("ファイルをオープンしました。\n");
	}

	for(i=0;fscanf(fp,"%c",&all[i])!=EOF;i++){ //改行も含め一つの配列に入れる
		} 

/*塩基配列（改行なし）を入れる*/
	i=0;
	j=0;	
	while(all[i]!='\0'){
		for(k=0;k<70;k++){ //1行70塩基
			data[i-j] = all[i];
			i++;
		}
		i++;
		j++;
	}
/*EcoRI*/
	printf("EcoRI\n");
	for(i=0,c=0;data[i+5] != '\0';i++,c++){
		if(data[i]=='G' && data[i+1]=='A' && data[i+2]=='A' && data[i+3]=='T' && data[i+4]=='T' && data[i+5]=='C'){
			l[i] = 'B';
		}
		else{
			l[i] = 'A';
		}
	}
	l[c]='A';
	l[c+1]='A';
	l[c+2]='A';
	l[c+3]='B'; //GAATTCのGの部分 と一番最後 がBであとは全てAの配列lができる
	l[c+4]='\0';
	for(i=0,d=0;data[i] !='\0';i++){
		d++;
		if(l[i]=='B'){
			printf("%d\n",d); //(Bの次にあるAから、)Bまでの個数を出力
			d = 0;
		}
	}


/*HindⅢ */	
	printf("\nHindⅢ\n");
	for(i=0,c=0;data[i+5] != '\0';i++,c++){
		if(data[i]=='A' && data[i+1]=='A' && data[i+2]=='G' && data[i+3]=='C' && data[i+4]=='T' && data[i+5]=='T'){
			l[i] = 'C';
		}
		else{
		}
	}
	l[c+3]='C'; //上のlにいくつかCが増える、以下同様
	for(i=0,d=0;data[i] !='\0';i++){
		d++;
		if(l[i]=='C'){
			printf("%d\n",d);
			d = 0;
		}
	}

/*BamHI*/	
	printf("\nBamHI\n");
	for(i=0,c=0;data[i+5] != '\0';i++,c++){
		if(data[i]=='G' && data[i+1]=='G' && data[i+2]=='A' && data[i+3]=='T' && data[i+4]=='C' && data[i+5]=='C'){
			l[i] = 'D';
		}
		else{
		}
	}
	l[c+3]='D';
	for(i=0,d=0;data[i] !='\0';i++){
		d++;
		if(l[i]=='D'){
			printf("%d\n",d);
			d = 0;
		}
	}

/*NotI*/
	printf("\nNotI\n");
	for(i=0,c=0;data[i+7] != '\0';i++,c++){
		if(data[i]=='G' && data[i+1]=='C' && data[i+2]=='G' && data[i+3]=='G' && data[i+4]=='C' && data[i+5]=='C' && data[i+6]=='G' && data[i+7]=='C'){
			l[i+1] = 'E';
		}
		else{
		}
	}
	l[c+5]='E';
	for(i=0,d=0;data[i] !='\0';i++){
		d++;
		if(l[i]=='E'){
			printf("%d\n",d);
			d = 0;
		}
	}
/*全部*/
	printf("\n全部\n");
	for(i=0,d=0;data[i] !='\0';i++){
		d++;
		if(l[i]!='A'&& l[i]!='\0'){
			printf("%d\n",d);
			d = 0;
		}
}	
	return 0;
}
